File: lib/position.rb
require_relative "../lib/pieces/chariot"
require_relative "../lib/pieces/horse"
require_relative "../lib/pieces/elephant"
require_relative "../lib/pieces/advisor"
require_relative "../lib/pieces/general"
require_relative "../lib/pieces/cannon"
require_relative "../lib/pieces/soldier"

class Position
  attr_reader :sides, :to_move

  def initialize
    empty_locations
    empty_sides
    get_moves
  end

  def place piece
    @locations[piece.x][piece.y] = piece
    @sides[piece.color] << piece
    self
  end

  def import hsh
    hsh[:pieces].each { |piece| place piece }
    @to_move = hsh[:to_move]
    get_moves
    self
  end

  def export
    { pieces: all_pieces, to_move: @to_move }
  end

  def all_pieces
    pieces = []
    @locations.keys.each do |x|
      @locations[x].keys.each do |y|
        pieces << @locations[x][y]
      end
    end
    pieces
  end

  def find hsh
    @locations[hsh[:x]][hsh[:y]]
  end

  def check?
    export_data = export
    export_data[:to_move] = not_to_move
    test_position = Position.new.import export_data
    general = find_general
    return true if general.sees_other? self
    general_coordinates = { x: general.x, y: general.y }
    test_position.all_pieces.each do |piece|
      return true if piece.moves.include? general_coordinates
    end
    false
  end

private
  
  def empty_locations
    @locations = {}
    (0..8).each do |i|
      @locations[i] = {}
    end
  end

  def empty_sides
    @sides = {red: [], black: []}
  end

  def get_moves
    all_pieces.each do |piece|
      piece.get_moves self if piece.color == to_move
    end
  end

  def not_to_move
    @to_move == :red ? :black : :red
  end

  def find_general
    all_pieces.each do |piece|
      return piece if piece.color == @to_move && piece.class == General
    end
  end
end

File: src/position.js
var Board = require('./board.js');

function Position() {
  var self = this;
  var toMove;

  function place(piece, coordinates) {
    this[coordinates] = piece;
    piece.square = this.BOARD[coordinates];
    return this;
  }

  function _import(position) {
    for (var key in position) {
      this.place(position[key], key)
    }
    this.toMove = position.toMove;
    return this;
  }

  function findGeneral() {
    for (var coordinates in self) {
      var piece = self[coordinates];
      if (piece.color === toMove && piece.type === 'General') {
        return coordinates;
      }
    }
  }

  function isCheck() {
    var general_location = findGeneral();
    for (var coordinates in self) {
      piece = self[coordinates];
      if (piece.color === toMove) {
        continue;
      }
      if (piece.getMoves(self).indexOf(general_location) !== -1) {
        return true;
      }
    }
    return false;
  }

  Object.defineProperties(this, {
    toMove: {
      get: function() { return toMove; },
      set: function(color) { toMove = color; },
      enumerable: false
    },
    place: { value: place, enumerable: false },
    'import': { value: _import, enumerable: false },
    isCheck: {
      get: function() { return isCheck(); },
      enumerable: false
    }
  });
}

Position.prototype = {
  BOARD: new Board()
};

module.exports = Position;
